### PR TITLE
api: translate constraint violation messages

### PR DESCRIPTION
**Pull request by BacLuc**
This PR implements #2902

Tasks in this PR:
- [x] Generate translation key from Constraint class name and expose in payload to frontend
- [x] Also expose the parameters for this translation to the frontend
- [x] Translate the messages into all configured locales and expose payload to frontend
- [x] Test that the frontend receives these translations and the translation key with parameters
- [x] Use the available translations already provided by symfony for the Constraints provided by symfony
- [x] Test support for self defined Translations for self defined Constraints
- [x] Test support for self defined Translations for variants like de_CH_scout
- [x] Integrate with lokalise (Lokalise can pull our files, so i will check that after merge): https://app.lokalise.com/project/177816575f917f180c4c07.19252326/?view=single&reference_lang_id=640&single_lang_id=10001&search=mail
- [x] Use Translation in Frontend
- [x] Support jsonapi and hydra format
- [x] ~~Maybe make support for jsonapi and hydra better by not using subclasses
(This is currently the only idea to share the functionality between the formats)~~

Not in this PR, will implement later
- switch to symfony translations for E-Mail

Sure not in this PR, and maybe we don't implement it at all:
- Translation of 400, 401, 403 and 404 Messages in the api
- Translation of 50x Messages in the api
- Translate constants which are exposed in the api (like CampCollaboration::status: invited, established...), because in some places we have concatenated values, which would increase the number of translations.

Is currently a draft, that you may look at it before everything is implemented.

Questions:

1. Why did we name our variants with dashes instead of underscores?
I know it with underscores (en_GB, en_US)
2. Do we really want to support jsonapi. Their error format is a little different than hydra and json+problem
https://github.com/api-platform/core/issues/3042
-> we don't support jsonapi (removed complicated code with: 415daa2c4e844089826d9714f20a522a18ef8fee)

### CODE REVIEW

File: api/translations/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Symfony translation files
+
+Docs: [https://symfony.com/doc/current/translation.html#message-format](https://symfony.com/doc/current/translation.html#message-format)
+
+We use the english message as the translation key because symfony does that as well for their validation messages.
```
Comment (usu) [inline api/translations/README.md:5]: I find this a rather odd design decision... 😵

File: api/translations/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+[https://github.com/symfony/validator/blob/v6.1.4/Resources/translations/validators.en.xlf](https://github.com/symfony/validator/blob/v6.1.4/Resources/translations/validators.en.xlf)
+We still add the english translation, that external translators could improve the maybe
```
Comment (usu) [inline api/translations/README.md:7]: Which might be Symfony's initial motivation to use the actual message as the translation key. 
But honestly, that's not really a situation I want to end up with and I'd use search/replace 100% times over starting to deviate the key & the message, if the message already was the key.

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php [hunk 3]
```diff
@@ -369,6 +369,52 @@ public function testCreateCampCollaborationValidatesMissingUserAndInviteEmail()
...
+        static::createClientWithCredentials()->request('POST', '/camp_collaborations', ['json' => $this->getExampleWritePayload([], ['user'])]);
+
+        $this->assertResponseStatusCodeSame(422);
+        $this->assertJsonContains([
+            'violations' => [
+                [
+                    'i18n' => [
+                        'key' => 'app.validator.asserteitherisnull',
+                        'parameters' => [
+                            '{{ other }}' => 'user',
```
Comment (usu) [inline api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: Do we really want the brackets?

Or would we better have it without:
`'other' => 'user'`

2nd might be easier to process by the client.

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php [hunk 4]
```diff
@@ -369,6 +369,52 @@ public function testCreateCampCollaborationValidatesMissingUserAndInviteEmail()
...
+                        ],
+                        'translations' => [
+                            'en' => 'Either this value or user should not be null.',
+                            'en_CH_scout' => 'Either this value or user should not be null.',
+                            'de' => 'Either this value or user should not be null.',
+                            'de_CH_scout' => 'Either this value or user should not be null.',
+                            'fr' => 'Either this value or user should not be null.',
+                            'fr_CH_scout' => 'Either this value or user should not be null.',
+                            'it' => 'Either this value or user should not be null.',
+                            'it_CH_scout' => 'Either this value or user should not be null.',
```
Comment (usu) [inline api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: Alternative: Only list the ones that have an actual translation and let the client decide on the fallback? Now it's a bit difficult for a client to know if a specific translation contains a real translation or only a fallback.

File: api/.env [hunk 5]
```diff
@@ -81,3 +81,5 @@ MAIL_FROM="info@ecamp3.ch"
 # https://www.google.com/recaptcha/admin
 RECAPTCHA_SECRET="disabled"
 ###< google/recaptcha ###
+
+TRANSLATE_ERRORS_TO_LOCALES="en,en_CH_scout,de,de_CH_scout,fr,fr_CH_scout,it,it_CH_scout"
```
Comment (usu) [inline api/.env]: Not sure if we need the scout versions at this stage already. I'd expect validation translation messages to be pretty generic, at least at the beginning.

**Review by usu — COMMENTED** [review]
General approach looks go to me. Some comments on the output format.

File: api/src/Serializer/Normalizer/TranslationConstraintViolationListNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Validator\ConstraintViolation;
+use Symfony\Component\Validator\ConstraintViolationList;
+
+class TranslationConstraintViolationListNormalizer implements NormalizerInterface, CacheableSupportsMethodInterface {
+    public const JSONAPI_FORMAT = ConstraintViolationListNormalizer::FORMAT;
+
+    public function __construct(
+        private readonly AbstractConstraintViolationListNormalizer $hydraNormalizer,
+        private readonly ConstraintViolationListNormalizer $jsonApiNormalizer,
```
Comment (BacLuc) [inline api/src/Serializer/Normalizer/TranslationConstraintViolationListNormalizer.php]: jsonApi weg

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php (hunk 3 above)
Comment (BacLuc) [reply api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: Falls ICU Message format standard -> so lassen
Sonst im backend die curly braces weg

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php (hunk 4 above)
Comment (BacLuc) [reply api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: scout varianten entfernen, nur das ausgeben, was wir haben.
Im backend nur variante ausgeben, wenn sie definiert ist.

File: api/.env (hunk 5 above)
Comment (BacLuc) [reply api/.env]: For the devel without dialects, but make sure that the dialects work.

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php (hunk 3 above)
Comment (BacLuc) [reply api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: I cannot derive the double curly braces from the ICU Message format, only single curly brace.
The double curly braces would occur with pluralization, but not with normal parameters.
-> use no curly braces at all.
done

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php (hunk 4 above)
Comment (BacLuc) [reply api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php]: done

File: api/.env (hunk 5 above)
Comment (BacLuc) [reply api/.env]: done

File: api/src/Serializer/Normalizer/TranslationConstraintViolationListNormalizer.php (hunk 6 above)
Comment (BacLuc) [reply api/src/Serializer/Normalizer/TranslationConstraintViolationListNormalizer.php]: done

File: api/translations/validators.de.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+Either this value or {{ other }} should not be null.: "Entweder dieser Wert oder {{ other }} müssen nicht null sein"
```
Comment (usu) [inline api/translations/validators.de.yml]: Tönt etwas holprig. Vorschlag:
```suggestion
Either this value or {{ other }} should not be null.: "Dieser Wert und {{ other }} dürfen nicht beide null sein."
```

File: frontend/src/helpers/__tests__/serverError.spec.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  name: 'ServerException',
+  response: {
+    status: 401,
+    data: {
+      detail: 'Unauthorized',
+    },
+  },
+}
+
+const validationError = {
```
Comment (usu) [inline frontend/src/helpers/__tests__/serverError.spec.js]: The selection of the wording makes it a bit hard to follow the tests. Should we be a bit more explicit in the message itself?

```javascript
const validationError = {
  name: 'ServerException',
  response: {
    status: 422,
    data: {
      violations: [
        {
          message: 'some message (untranslated)',
          propertyPath: 'parent.property',
          i18n: {
            key: 'some.key',
            parameters: {},
            translations: {
              de: 'some message (de translated)',
            },
          },
        },
        {
          message: 'another message',
          propertyPath: 'parent.property',
          i18n: {
            key: 'some.other.key',
            parameters: {},
            translations: {
              de: 'another message (de translated)',
            },
          },
        },
        {
          message: 'yet another message',
          propertyPath: 'property2',
          i18n: {
            key: 'yet.other.key',
            parameters: {},
            translations: {
              fr: 'yet another message (fr translated)',
              fr_CH_scout: 'yet another message (fr_CH_scout translated)',
              en: 'yet another message (en translated)',
            },
          },
        },
      ],
    },
  },
}
```

**Review by usu — APPROVED** [review]
Nice. Concept + generally looks good to me. Minor comments/improvements as below

File: api/translations/validators.de.yml (hunk 7 above)
Comment (BacLuc) [reply api/translations/validators.de.yml]: done, see the diff here: https://github.com/ecamp/ecamp3/compare/0fe9d6079a80c42a07f3670e3b49ffdff8d72d63..2ad763ae2cd2528bbfc798782f5bf62a3e476208

File: frontend/src/helpers/__tests__/serverError.spec.js (hunk 8 above)
Comment (BacLuc) [reply frontend/src/helpers/__tests__/serverError.spec.js]: Went even a step further and made the tests independent of each other.
See the diff here: https://github.com/ecamp/ecamp3/compare/0fe9d6079a80c42a07f3670e3b49ffdff8d72d63..2ad763ae2cd2528bbfc798782f5bf62a3e476208

File: api/translations/validators.de.yml (hunk 7 above)
Comment (carlobeltrame) [reply api/translations/validators.de.yml]: So ist es aber ungenau. Es dürfen nicht beide `null` sein, aber es dürfen auch nicht beide ausgefüllt sein.

Comment (usu) [reply api/translations/validators.de.yml]: True, `AssertEitherIsNull` validates both cases, but works with 2 different validation messages.
```php
    public string $messageBothNull = 'Either this value or {{ other }} should not be null.';
    public string $messageNoneNull = 'Either this value or {{ other }} should be null.';
```

The first one I'd still translate as above.

The second one is actually not translated at the moment. I'd propose "Entweder dieser Wert oder {{ other }} muss null sein."

File: api/.env.test [hunk 9]
```diff
@@ -6,3 +6,4 @@ PANTHER_APP_ENV=panther
 PANTHER_ERROR_SCREENSHOT_DIR=./var/error-screenshots
 API_DOMAIN=example.com
 COOKIE_PREFIX=example_com_
+TRANSLATE_ERRORS_TO_LOCALES="en,en_CH_scout,de,de_CH_scout,fr,fr_CH_scout,it,it_CH_scout"
```
Comment (carlobeltrame) [inline api/.env.test:9]: The syntax with underscores was invented for Java I think. The syntax with dashes is the [IETF language tag format](https://en.wikipedia.org/wiki/IETF_language_tag). On the web, the IETF dash syntax is more common (e.g. Accept-Language header, html lang="" attribute). Do you need to use underscores here, or could we use dashes as well?

File: api/config/services.yaml [hunk 10]
```diff
@@ -131,6 +136,13 @@ services:
             - "%env(FRONTEND_BASE_URL)%"
             - "%env(MAIL_FROM)%"
 
+    App\Service\TranslateToAllLocalesService:
+        public: true
+        arguments:
+            - '@translator'
+            - '@translator'
```
Comment (carlobeltrame) [inline api/config/services.yaml:143]: Is twice the same argument really correct here? Does that take the type annotation into account in addition to the alias which you write here?

File: frontend/src/helpers/serverError.js [hunk 11]
```diff
@@ -12,27 +14,50 @@ const serverErrorToString = (error) => {
   return error.message
 }
 
+function getApiTranslation(locale, violation) {
+  if (!locale) {
+    return null
+  }
+  const matchingTranslation = [locale, ...fallbackLocalesFor(locale)]
+    .map((locale) => violation.i18n?.translations[locale])
```
Comment (carlobeltrame) [inline frontend/src/helpers/serverError.js]: If `i18n` is `undefined`, then `i18n?.translations` is silently `undefined` as well, and the array access will throw an error. You can use the optional chaining array access operator:
```suggestion
    .map((locale) => violation.i18n?.translations?.[locale])
```
Looks strange, but works.

**Review by carlobeltrame — APPROVED** [review]

File: api/config/services.yaml (hunk 10 above)
Comment (BacLuc) [reply api/config/services.yaml:143]: Symfony translations does not expose the TranslatorBagInterface as service, but the implementation of TranslatorInterface offered by Symfony translations also implements that interface.
Because Union Types did not work well with the DI last time, i did not try it this time.
Yes, it's on purpose. The same class, but as the implementation of 2 different interfaces.

File: frontend/src/helpers/serverError.js (hunk 11 above)
Comment (BacLuc) [reply frontend/src/helpers/serverError.js]: It's our api, so i18n always has this key, even when the env variable TRANSLATE_ERRORS_TO_LOCALES is empty.
Will fix

File: api/translations/validators.de.yml (hunk 7 above)
Comment (BacLuc) [reply api/translations/validators.de.yml]: I propose we add the translations themselves in a later PR.

File: api/.env.test (hunk 9 above)
Comment (BacLuc) [reply api/.env.test:9]: I tested it quickly, api platform does not support dashes...

File: frontend/src/helpers/serverError.js (hunk 11 above)
Comment (BacLuc) [reply frontend/src/helpers/serverError.js]: done

File: api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php [hunk 12]
```diff
@@ -369,6 +369,44 @@ public function testCreateCampCollaborationValidatesMissingUserAndInviteEmail()
         ]);
     }
 
+    public function testCreateCampCollaborationValidatesMissingUserAndInviteEmailAndIncludesTranslationInfo() {
```
Comment (BacLuc) [inline api/tests/Api/CampCollaborations/CreateCampCollaborationTest.php:372]: just as ref
